### DISCUSSION TITLE
Failed to open TCP connection to localhost:3036

### DISCUSSION THREAD

**Discussion by johnvictorfs** [ElMassimo/vite_ruby#89] [Q&A] [answered]
### Description 📖

Hello, I've been trying to migrate a Rails app from webpacker to vite-rails, and while everything works with the `bin/vite build` command (or the automatic build when the dev server is not running) I cannot get Rails to connect correctly to the vite dev server. The error logs and other relevant logs are down below.

I unfortunately could not reproduce this error on a new project so it is likely to be an issue with something else on my app that's not letting Rails connect to the vite dev server for some reason. But other connections are working fine (like to `webpack-dev-server` or normal external API requests).

While this may not be an issue with vite_ruby/vite_rails itself, would anyone know what could possibly be going wrong with my setup? I have tried not running most extra middleware and most things from the initializers folders that's not default by Rails, also with no success.

### Reproduction 🐞

<details>
<summary>Vite Ruby Info</summary>

_Run `bin/rake vite:info` and provide the output:_

```
bin/vite present?: true
vite_ruby: 1.2.11
vite_rails: 2.0.12
rails: 6.1.3.1
node: v14.16.1
npm: 6.14.12
yarn: 1.22.10
pnpm: 
ruby: ruby 2.6.6p146 (2020-03-31 revision 67876) [x86_64-linux]

installed packages:
/home/***/projects/***
├── vite@2.3.6 
└── vite-plugin-ruby@2.0.4 
```
</details>

### Logs 📜

Vite dev server itself seems to run just fine with seemingly nothing wrong on logs

_If not providing a reproduction:_

<details>
<summary>Output</summary>

_Run `DEBUG=vite-plugin-ruby:* bin/vite dev` or `DEBUG=vite-plugin-ruby:* bin/vite build` and provide the output:_

```
 vite-plugin-ruby:config {
  vite-plugin-ruby:config   base: '/vite-dev/',
  vite-plugin-ruby:config   build: {
  vite-plugin-ruby:config     emptyOutDir: true,
  vite-plugin-ruby:config     sourcemap: false,
  vite-plugin-ruby:config     assetsDir: 'assets',
  vite-plugin-ruby:config     manifest: true,
  vite-plugin-ruby:config     outDir: '../../../public/vite-dev',
  vite-plugin-ruby:config     rollupOptions: { input: [Object], output: [Object] }
  vite-plugin-ruby:config   },
  vite-plugin-ruby:config   root: '/home/***/projects/***/app/javascript/entrypoints',
  vite-plugin-ruby:config   server: {
  vite-plugin-ruby:config     host: 'localhost',
  vite-plugin-ruby:config     https: false,
  vite-plugin-ruby:config     port: 3036,
  vite-plugin-ruby:config     strictPort: true,
  vite-plugin-ruby:config     fsServe: { root: '/home/***/projects/***' }
  vite-plugin-ruby:config   },
  vite-plugin-ruby:config   entrypoints: {
  vite-plugin-ruby:config     App: '/home/***/projects/***/app/javascript/entrypoints/App.jsx',
  vite-plugin-ruby:config     Routes: '/home/***/projects/***/app/javascript/entrypoints/Routes.jsx',
  vite-plugin-ruby:config     application: '/home/***/projects/***/app/javascript/entrypoints/application.js',
  vite-plugin-ruby:config     wdyr: '/home/***/projects/***/app/javascript/entrypoints/wdyr.js',
  vite-plugin-ruby:config ...
  vite-plugin-ruby:config   }
  vite-plugin-ruby:config } +0ms

  vite v2.3.6 dev server running at:

  > Local: http://localhost:3036/vite-dev/
  > Network: use `--host` to expose

  ready in 352ms.
```

</details>

<details>
<summary>Rails error logs</summary>

```
2021-06-08 19:48:20 -0300: Rack app error handling request { GET /vite-dev/App.jsx }
#<Errno::ECONNREFUSED: Failed to open TCP connection to localhost:3036 (Connection refused - connect(2) for "::1" port 3036)>
/home/***/.rbenv/versions/2.6.6/lib/ruby/2.6.0/net/http.rb:949:in `rescue in block in connect'
/home/***/.rbenv/versions/2.6.6/lib/ruby/2.6.0/net/http.rb:946:in `block in connect'
/home/***/.rbenv/versions/2.6.6/lib/ruby/2.6.0/timeout.rb:93:in `block in timeout'
/home/***/.rbenv/versions/2.6.6/lib/ruby/2.6.0/timeout.rb:103:in `timeout'
/home/***/.rbenv/versions/2.6.6/lib/ruby/2.6.0/net/http.rb:945:in `connect'
/home/***/.rbenv/versions/2.6.6/lib/ruby/2.6.0/net/http.rb:930:in `do_start'
/home/***/.rbenv/versions/2.6.6/lib/ruby/2.6.0/net/http.rb:925:in `start'
/home/***/.ruby/gems/rack-proxy-0.6.5/lib/rack/http_streaming_response.rb:71:in `session'
/home/***/.ruby/gems/rack-proxy-0.6.5/lib/rack/http_streaming_response.rb:60:in `response'
/home/***/.ruby/gems/rack-proxy-0.6.5/lib/rack/http_streaming_response.rb:29:in `headers'
/home/***/.ruby/gems/rack-proxy-0.6.5/lib/rack/proxy.rb:120:in `perform_request'
/home/***/.ruby/gems/vite_ruby-1.2.11/lib/vite_ruby/dev_server_proxy.rb:21:in `perform_request'
/home/***/.ruby/gems/rack-proxy-0.6.5/lib/rack/proxy.rb:57:in `call'
/home/***/.ruby/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101:in `call'
/home/***/.ruby/gems/railties-6.1.3.1/lib/rails/engine.rb:539:in `call'
/home/***/.ruby/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101:in `call'
/home/***/.ruby/gems/puma-3.12.6/lib/puma/configuration.rb:227:in `call'
/home/***/.ruby/gems/puma-3.12.6/lib/puma/server.rb:706:in `handle_request'
/home/***/.ruby/gems/puma-3.12.6/lib/puma/server.rb:476:in `process_client'
/home/***/.ruby/gems/puma-3.12.6/lib/puma/server.rb:334:in `block in run'
/home/***/.ruby/gems/puma-3.12.6/lib/puma/thread_pool.rb:135:in `block in spawn_thread'
2021-06-08 19:48:20 -0300: Rack app error handling request { GET /vite-dev/@vite/client }
#<Errno::ECONNREFUSED: Failed to open TCP connection to localhost:3036 (Connection refused - connect(2) for "::1" port 3036)>
/home/***/.rbenv/versions/2.6.6/lib/ruby/2.6.0/net/http.rb:949:in `rescue in block in connect'
/home/***/.rbenv/versions/2.6.6/lib/ruby/2.6.0/net/http.rb:946:in `block in connect'
/home/***/.rbenv/versions/2.6.6/lib/ruby/2.6.0/timeout.rb:93:in `block in timeout'
/home/***/.rbenv/versions/2.6.6/lib/ruby/2.6.0/timeout.rb:103:in `timeout'
/home/***/.rbenv/versions/2.6.6/lib/ruby/2.6.0/net/http.rb:945:in `connect'
/home/***/.rbenv/versions/2.6.6/lib/ruby/2.6.0/net/http.rb:930:in `do_start'
/home/***/.rbenv/versions/2.6.6/lib/ruby/2.6.0/net/http.rb:925:in `start'
/home/***/.ruby/gems/rack-proxy-0.6.5/lib/rack/http_streaming_response.rb:71:in `session'
/home/***/.ruby/gems/rack-proxy-0.6.5/lib/rack/http_streaming_response.rb:60:in `response'
/home/***/.ruby/gems/rack-proxy-0.6.5/lib/rack/http_streaming_response.rb:29:in `headers'
/home/***/.ruby/gems/rack-proxy-0.6.5/lib/rack/proxy.rb:120:in `perform_request'
/home/***/.ruby/gems/vite_ruby-1.2.11/lib/vite_ruby/dev_server_proxy.rb:21:in `perform_request'
/home/***/.ruby/gems/rack-proxy-0.6.5/lib/rack/proxy.rb:57:in `call'
/home/***/.ruby/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101:in `call'
/home/***/.ruby/gems/railties-6.1.3.1/lib/rails/engine.rb:539:in `call'
/home/***/.ruby/gems/newrelic_rpm-7.0.0/lib/new_relic/agent/instrumentation/middleware_tracing.rb:101:in `call'
/home/***/.ruby/gems/puma-3.12.6/lib/puma/configuration.rb:227:in `call'
/home/***/.ruby/gems/puma-3.12.6/lib/puma/server.rb:706:in `handle_request'
/home/***/.ruby/gems/puma-3.12.6/lib/puma/server.rb:476:in `process_client'
/home/***/.ruby/gems/puma-3.12.6/lib/puma/server.rb:334:in `block in run'
/home/***/.ruby/gems/puma-3.12.6/lib/puma/thread_pool.rb:135:in `block in spawn_thread'
```
</details>

<details>
<summary>vite.config.ts and vite.json</summary>

```ts
import { resolve } from 'path'
import { defineConfig } from 'vite'
import RubyPlugin from 'vite-plugin-ruby'

export default defineConfig({
  plugins: [RubyPlugin()],
  resolve: {
    alias: {
      '@': resolve(__dirname, 'app/javascript/entrypoints'),
      '@assets': resolve(__dirname, 'app/assets')
    }
  }
})
```

```json
{
  "all": {
    "sourceCodeDir": "app/javascript",
    "watchAdditionalPaths": []
  },
  "development": {
    "publicOutputDir": "vite-dev",
    "port": 3036,
    "autoBuild": true, // I have also tried using not using autoBuild
    "devServerConnectTimeout": 3 // I have tried a lot of different timeouts, and using the default one, with no success 
  },
  "test": {
    "autoBuild": true,
    "publicOutputDir": "vite-test"
  }
}
```

</details>

### Screenshots 📷

_Provide console or browser screenshots of the problem_.

![image](https://user-images.githubusercontent.com/37747572/121268576-9d050900-c894-11eb-9c4d-4e7313d432a5.png)

![image](https://user-images.githubusercontent.com/37747572/121268604-aaba8e80-c894-11eb-8678-f64f7e04dc2b.png)

![image](https://user-images.githubusercontent.com/37747572/121268815-0d138f00-c895-11eb-8c25-74fd07e344de.png)

- Auto-build with vite server not running works fine
  ![image](https://user-images.githubusercontent.com/37747572/121269121-a04cc480-c895-11eb-82b9-ba1456df6a64.png)

**Comment by ElMassimo** [comment]
Hi John Victor!

Nothing unusual in the config, except that it might be better to be explicit with the `@` alias by using `@/` instead:

```js
    alias: {
      '@/': `${resolve(__dirname, 'app/javascript/entrypoints')}/`,
```

Otherwise it could interfere with certain imports such as `@vite/client` or `@react-refresh`.

<hr/>

Based on the Rails error logs, it seems that the Vite dev server is not reachable. You'll want to [verify this from a Rails console](https://vite-ruby.netlify.app/guide/troubleshooting.html#build-is-triggered-when-the-dev-server-is-running), by using `ViteRuby.instance.dev_server_running?`.

- If it returns `false`, then the problem is likely to be a [network issue](https://github.com/ElMassimo/vite_ruby/discussions/33)
- Otherwise, the problem might be related with how the files are being served by Vite

What is the response if you `curl` `http://localhost:3036/vite-dev/App.jsx` or `http://localhost:3036/vite-dev/@vite/client` in your browser?

**Reply by ElMassimo** [reply]
PS: @johnvictorfs Moved it to _Discussions_, since it's likely to be a problem in how the project is configured.

**Reply by johnvictorfs** [reply]
`ViteRuby.instance.dev_server_running?` does return `true`

![image](https://user-images.githubusercontent.com/37747572/121271674-8a8dce00-c89a-11eb-867b-939f242fe06e.png)

- `http://localhost:3036/vite-dev/App.jsx` returns a page with the same code as in the original `App.jsx` file, and `http://localhost:3036/vite-dev/@vite/client` just returns a blank page

![image](https://user-images.githubusercontent.com/37747572/121271714-a42f1580-c89a-11eb-8f6d-38fa604dfe5a.png)

![image](https://user-images.githubusercontent.com/37747572/121271751-ba3cd600-c89a-11eb-8a57-697038d752a2.png)

**Reply by ElMassimo** [reply]
@johnvictorfs Let's try a few things.

- Can you check the `Content-Type` header when you request the `App.jsx` file to make sure it's `application/javascript`?
- Try running `bin/vite dev --debug` so that Vite prints more information about the incoming requests, to see if it logs anything when Rails attempts to proxy a request.

If you are online, you can join the [Vite.js Discord channel](https://discord.gg/hDNyAuM3) and I'll help you debug it.

**Reply by johnvictorfs** [reply]
`Content-Type` seems correct for `App.jsx`
  
![image](https://user-images.githubusercontent.com/37747572/121272479-73e87680-c89c-11eb-8e2b-c4e6f2c7ef07.png)

- With the `--debug` flag there are no extra messages when accessing any rails pages, those are the messages when I access the files manually:

```
vite v2.3.6 dev server running at:

  > Local: http://localhost:3036/vite-dev/
  > Network: use `--host` to expose

  ready in 716ms.

  vite:resolve 2ms   /App.jsx -> /home/***/projects/***/app/javascript/entrypoints/App.jsx +0ms
  vite:time 8ms   /App.jsx +0ms
  vite:resolve 1ms   /home/***/projects/***/node_modules/vite/dist/client/client -> /home/***/projects/***/node_modules/vite/dist/client/client.js +18s
  vite:resolve 2ms   /@vite/client -> /home/***/projects/***/node_modules/vite/dist/client/client.js +0ms
  vite:spa-fallback Rewriting GET /home/***/projects/***/node_modules/vite/dist/client/client to /index.html +0ms
  vite:time 5ms   /index.html +18s
```

Yeah, I can join the Discord in just a bit, thanks

**Comment by ElMassimo** [comment, marked as answer]
The problem is related with a [recent change in Vite](https://github.com/vitejs/vite/pull/2977/files#diff-35ba301b85014a4bfaa9cad2d8e7eafa41c4e8c2ddd5c193182241d9a1542082R45-R47), which affects systems where `localhost` defaults to `::1`(IPv6).

Configuring [`host`](https://vite-ruby.netlify.app/config/#host) explicitly to `127.0.0.1` should fix the problem.

```json
  "development": {
    "autoBuild": true,
    "host": "127.0.0.1",
    "publicOutputDir": "vite-dev",
    "port": 3036
  },
```

I will evaluate changing the default from `localhost` to `127.0.0.1` since at the moment [Vite performs that conversion](https://github.com/vitejs/vite/pull/2977/files#diff-35ba301b85014a4bfaa9cad2d8e7eafa41c4e8c2ddd5c193182241d9a1542082R45-R47) and causes this issue.

**Reply by KonnorRogers** [reply]
@ElMassimo I just ran into this when upgrading. is there any reason this shouldnt be the default?

**Reply by ElMassimo** [reply]
Hi Konnor! Only reason I can think of now is backwards compatibility, since the Ruby gem uses `localhost` and a mismatch could potentially cause problems (don't have any tests around host resolution).

Happy to explore that and make it the default in a next version, contributions are welcome!

**Reply by KonnorRogers** [reply]
Hmmm...yea testing host resolution is kind of a pain. Im tempted to think it's a documentation change (it's technically in the docs, but it's kind of tucked away under 404 errors, but the error that happens is far more arcane. On MacOS it sends itself into an infinite loop trying to find the file and then eventually I have to manually kill the Rails dev server. I don't even get the typical 404 in the network tab. I'll make a documentation PR to make a note of the error I saw.

**Reply by ElMassimo** [reply]
Let's explore having this by default in Vite Ruby, I think the only downside would be for Docker setups, but that can be documented and it's less time consuming to debug than this.

**Reply by KonnorRogers** [reply]
If Docker is a concern, would there be any harm in defaulting it to `0.0.0.0`

**Reply by ElMassimo** [reply]
`0.0.0.0` was the default in Vite, but since it provides transparent access to the filesystem, it's not a secure default.

Users now have to explicitly open the host if they want to allow external connections.

We can add a Docker configuration section in the Vite Ruby docs (instead of [my outdated example](https://github.com/ElMassimo/vite_rails_docker_example)).

**Reply by eadams17** [reply]
Hi, @ElMassimo! Thank you for all of your work on this project. I think I am encountering a similar issue, but I can't seem to figure out what I might be missing. I am also working on migrating our React/Rails app from using Webpacker to Vite. This is admittedly out of my usual wheelhouse as I haven't worked much in devops, so any help would be appreciated. I am also able to run the local Vite dev server with no issues, but I am getting a HTTP 404 error when trying to view http://127.0.0.1:3036/vite-dev/ (and there are no errors in the rails logs).

- Vite Ruby Info
```
bin/vite present?: true
vite_ruby: 3.2.3
vite_rails: 3.0.12
rails: 6.1.6
node: v16.13.1
npm: 8.1.2
yarn: 
pnpm: 
ruby: ruby 2.7.6p219 (2022-04-12 revision c9c2245c0a) [arm64-darwin20]

installed packages:
Empirical-Core@ /Users/ericadams/Documents/GitHub/Empirical-Core/services/QuillLMS
├─┬ vite-plugin-ruby@3.1.2
│ └── vite@3.0.9 deduped
└── vite@3.0.9
```

- `DEBUG=vite-plugin-ruby:* bin/vite dev` output
```
  vite-plugin-ruby:config {
  vite-plugin-ruby:config   base: '/vite-dev/',
  vite-plugin-ruby:config   build: {
  vite-plugin-ruby:config     emptyOutDir: true,
  vite-plugin-ruby:config     sourcemap: false,
  vite-plugin-ruby:config     assetsDir: 'assets',
  vite-plugin-ruby:config     manifest: true,
  vite-plugin-ruby:config     outDir: '../../public/vite-dev',
  vite-plugin-ruby:config     rollupOptions: { input: [Object], output: [Object] }
  vite-plugin-ruby:config   },
  vite-plugin-ruby:config   envDir: '/Users/ericadams/Documents/GitHub/Empirical-Core/services/QuillLMS',
  vite-plugin-ruby:config   root: '/Users/ericadams/Documents/GitHub/Empirical-Core/services/QuillLMS/app/javascript',
  vite-plugin-ruby:config   server: {
  vite-plugin-ruby:config     fs: { allow: [Array], strict: true },
  vite-plugin-ruby:config     host: '127.0.0.1',
  vite-plugin-ruby:config     https: false,
  vite-plugin-ruby:config     port: 3036,
  vite-plugin-ruby:config     strictPort: true,
  vite-plugin-ruby:config     hmr: { clientPort: 3036 }
  vite-plugin-ruby:config   },
  vite-plugin-ruby:config   entrypoints: {
  vite-plugin-ruby:config     'entrypoints/application.js': '/Users/ericadams/Documents/GitHub/Empirical-Core/services/QuillLMS/app/javascript/entrypoints/application.js'
  vite-plugin-ruby:config   }
  vite-plugin-ruby:config } +0ms
  ```

-  Rails log
```
=> Booting Puma
=> Rails 6.1.6 application starting in development 
=> Run `bin/rails server --help` for more startup options
[22085] Puma starting in cluster mode...
[22085] * Puma version: 5.6.4 (ruby 2.7.6-p219) ("Birdie's Version")
[22085] *  Min threads: 1
[22085] *  Max threads: 10
[22085] *  Environment: development
[22085] *   Master PID: 22085
[22085] *      Workers: 3
[22085] *     Restarts: (✔) hot (✖) phased
[22085] * Preloading application
[22085] * Listening on http://127.0.0.1:3000
[22085] * Listening on http://[::1]:3000
[22085] Use Ctrl-C to stop
DEPRECATION WARNING: [] is deprecated and will be removed from Rails 7.0 (Use configs_for) (called from block (2 levels) in _load_from at config/puma.rb:15)
DEPRECATION WARNING: [] is deprecated and will be removed from Rails 7.0 (Use configs_for) (called from block (2 levels) in _load_from at config/puma.rb:15)
DEPRECATION WARNING: [] is deprecated and will be removed from Rails 7.0 (Use configs_for) (called from block (2 levels) in _load_from at config/puma.rb:15)
[22108] WARNING hook before_worker_boot failed with exception (FrozenError) can't modify frozen Hash: {:adapter=>"postgresql", :encoding=>"unicode", :pool=>5, :prepared_statements=>false, :timeout=>5000, :host=>"localhost", :port=>5432, :database=>"emp_gr_development"}
[22085] - Worker 1 (PID: 22108) booted in 0.0s, phase: 0
[22109] WARNING hook before_worker_boot failed with exception (FrozenError) can't modify frozen Hash: {:adapter=>"postgresql", :encoding=>"unicode", :pool=>5, :prepared_statements=>false, :timeout=>5000, :host=>"localhost", :port=>5432, :database=>"emp_gr_development"}
[22085] - Worker 2 (PID: 22109) booted in 0.0s, phase: 0
[22107] WARNING hook before_worker_boot failed with exception (FrozenError) can't modify frozen Hash: {:adapter=>"postgresql", :encoding=>"unicode", :pool=>5, :prepared_statements=>false, :timeout=>5000, :host=>"localhost", :port=>5432, :database=>"emp_gr_development"}
[22085] - Worker 0 (PID: 22107) booted in 0.0s, phase: 0
```

- vite.config.ts and vite.json
```
import { resolve } from 'path'
import { defineConfig } from 'vite'
import RubyPlugin from 'vite-plugin-ruby'

export default defineConfig({
  plugins: [RubyPlugin()],
  resolve: {
    alias: {
      '@': resolve(__dirname, 'app/javascript/entrypoints'),
      '@assets': resolve(__dirname, 'app/assets')
    }
  }
})
```

```
{
  "all": {
    "sourceCodeDir": "app/javascript",
    "watchAdditionalPaths": []
  },
  "development": {
    "autoBuild": true,
    "host": "127.0.0.1",
    "publicOutputDir": "vite-dev",
    "port": 3036
  },
  "test": {
    "autoBuild": true,
    "publicOutputDir": "vite-test",
    "port": 3037
  }
}
```

- Screenshots
![Screen Shot 2022-08-31 at 3 17 51 PM](https://user-images.githubusercontent.com/25959584/187764428-b4f27984-3b54-4ace-92c9-168baa4b7e77.png)
![Screen Shot 2022-08-31 at 3 18 05 PM](https://user-images.githubusercontent.com/25959584/187764434-06dbae65-991c-41ee-a0db-061d82b0847c.png)
<img width="1301" alt="Screen Shot 2022-08-31 at 3 16 35 PM" src="https://user-images.githubusercontent.com/25959584/187764442-ecde9e07-58c1-4351-b6e4-510ca038e10f.png">
<img width="1440" alt="Screen Shot 2022-08-31 at 3 18 36 PM" src="https://user-images.githubusercontent.com/25959584/187764455-a2d4797f-2f8d-41de-b614-b30bbd348abc.png">

This is the [link to our repo](https://github.com/empirical-org/Empirical-Core). Any help would be much appreciated!

**Reply by ElMassimo** [reply]
Hi Eric! @eadams17

> I am getting a HTTP 404 error when trying to view `http://127.0.0.1:3036/vite-dev/`

Unless you have an `index.html` file in your source code dir, there's nothing Vite would serve in that path, so getting a 404 is the expected behavior.

Most of the time, you wouldn't access the Vite dev server directly, and instead would visit your _Rails_ server as usual during development.
